Sources/SwiftProtobuf/AnyMessageStorage.swift

```diff
@@ -236,31 +243,41 @@ extension AnyMessageStorage {
       return false
     }
 
-    // If we have both data's, compare those.
-    // The best option is to decode and compare the messages; this
-    // insulates us from variations in serialization details.  For
-    // example, one Any might hold protobuf binary bytes from one
-    // language implementation and the other from another language
-    // implementation.  But of course this only works if we
-    // actually know the message type.
-    //if let myMessage = _message {
-    //    if let otherMessage = other._message {
-    //        ... compare them directly
-    //    } else {
-    //        ... try to decode other and compare
-    //    }
-    //} else if let otherMessage = other._message {
-    //    ... try to decode ourselves and compare
-    //} else {
-    //    ... try to decode both and compare
-    //}
-    // If we don't know the message type, we have few options:
-    // If we were both deserialized from proto, compare the binary value:
-    // If we were both deserialized from JSON, compare content of the JSON?
+    // Since the library does lazy Any decode, equality is a very hard problem.
+    // It things exactly match, that's pretty easy, otherwise, one ends up having
+    // to error on saying they aren't equal.
+    //
+    // The best option would be to have Message forms and compare those, as that
+    // removes issues like map<> serialization order, some other protocol buffer
+    // implementation details/bugs around serialized form order, etc.; but that
+    // would also greatly slow down equality tests.
+    //
+    // Do our best to compare what is present have...
+
+    // If both have messages, check if they are the same.
+    if let myMsg = _message, let otherMsg = other._message, type(of: myMsg) == type(of: otherMsg) {
+      // Since the messages are known to be same type, we can claim both equal and
+      // not equal based on the equality comparison.
+      return myMsg.isEqualTo(message: otherMsg)
+    }
+
+    // If both have serialized data, and they exactly match; the messages are equal.
+    // Because there could be map in the message, the fact that the data isn't the
+    // same doesn't always mean the messages aren't equal.
```

Comment:
There are more issues with binary serialization than just `map`.  Protobuf recommends (but does not require) fields be in numeric order.  Unknown fields are routinely re-ordered as they get decoded/recoded.  Even repeated array fields can get reordered in some cases (the unknown enum problem).  The same message could get decoded/recoded by different systems and end up with very different binary forms for any of these reasons.